crates/partition-store/src/persisted_lsn_tracking.rs

```diff
@@ -0,0 +1,238 @@
+            }
+        }
+        true
+    }
+
+    fn finish(&mut self, properties: &mut HashMap<CString, CString>) -> bool {
+        for (partition_id, lsn) in &self.applied_lsns {
+            properties.insert(
+                CString::new(applied_lsn_property_name(*partition_id)).unwrap(),
+                CString::new(lsn.as_u64().to_string()).unwrap(),
+            );
+        }
+        true
+    }
+
+    fn get_readable_properties(&self) -> HashMap<CString, CString> {
+        let mut properties = HashMap::new();
+        for (partition_id, lsn) in &self.applied_lsns {
+            properties.insert(
+                CString::new(applied_lsn_property_name(*partition_id)).unwrap(),
+                CString::new(lsn.as_u64().to_string()).unwrap(),
+            );
+        }
+        properties
+    }
+
+    fn name(&self) -> &CStr {
+        c"LsnTrackingCollector"
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct LatestAppliedLsnCollectorFactory {}
+
+impl TablePropertiesCollectorFactory for LatestAppliedLsnCollectorFactory {
+    type Collector = LatestAppliedLsnCollector;
+
+    fn create(
+        &mut self,
+        _context: rocksdb::table_properties::TablePropertiesCollectorContext,
+    ) -> LatestAppliedLsnCollector {
+        LatestAppliedLsnCollector::new()
+    }
+
+    fn name(&self) -> &CStr {
+        c"AppliedLsnTrackingCollectorFactory"
+    }
+}
+
+pub struct PersistedLsnEventListener {
+    pub(crate) persisted_lsn_tx: mpsc::Sender<(PartitionId, Lsn)>,
+}
+
+impl EventListener for PersistedLsnEventListener {
+    fn on_flush_completed(&self, flush_job_info: FlushJobInfo) {
+        if let Some(id_str) = flush_job_info.cf_name.strip_prefix("data-") {
+            let Ok(id) = id_str.parse::<u16>() else {
+                warn!(
+                    "Failed to parse partition id from cf_name: {}",
+                    flush_job_info.cf_name
+                );
+                return;
+            };
+
+            let partition_id = PartitionId::from(id);
+            let key = applied_lsn_property_name(partition_id);
+            if let Some(applied_lsn) = flush_job_info.get_user_collected_property(&key) {
+                match applied_lsn.to_str().expect("valid string").parse::<u64>() {
+                    Ok(persisted_lsn) => {
+                        // ignore shutdown errors
+                        let _ = self
+                            .persisted_lsn_tx
+                            .blocking_send((partition_id, persisted_lsn.into()));
+                    }
+                    Err(err) => warn!(
+                        key,
+                        "Failed to parse applied LSN from table property: {}", err
+                    ),
+                };
+            }
+        }
+    }
+}
+
+/// Custom table property name for Restate applied LSN
+#[inline]
+fn applied_lsn_property_name(partition_id: PartitionId) -> String {
+    format!("restate.partition_{}.applied_lsn", partition_id)
```

Comment:
Overkill? I don't want to make it too magical; we can probably change this easily for as long as we only ever read the property once on flush.

Reply:
This is indeed a possible future improvement. Once we have measured that it is noticeable.